src/Voron/Data/CompactTrees/CompactTree.cs

```diff
@@ -34,59 +34,518 @@ private struct IteratorCursorState
+
+            public long Dictionary { get; private set; }
+
+            private const int Invalid = -1;
+
+            public EncodedKey(CompactTree tree)
+            {
+                Owner = tree;
+                Dictionary = Invalid;
+                _currentKeyIdx = Invalid;
+                _decodedKeyIdx = Invalid;
+                _keyMappingCurrent = Invalid;
+
+                _storageScope = Owner.Llt.Allocator.Allocate(Constants.CompactTree.MaximumKeySize * 2, out _storage);
+                _currentPtr = _storage.Ptr; 
+                _currentEndPtr = _currentPtr + Constants.CompactTree.MaximumKeySize * 2;
+            }
+
+            public EncodedKey(EncodedKey key)
+            {
+                // This operation is essentially a simplifier Set where we would just copy
+                // the entire structure of the key through memory copy. This behaves like a
+                // deep clone.
+
+                Owner = key.Owner;
+                Dictionary = key.Dictionary;
+                _currentKeyIdx = key._currentKeyIdx;
+                _decodedKeyIdx = key._decodedKeyIdx;
+                _keyMappingCurrent = key._keyMappingCurrent;
+
+                _storageScope = Owner.Llt.Allocator.Allocate(key._storage.Size, out _storage);
+                _currentPtr = _storage.Ptr;
+                long size = (key._currentEndPtr - key._storage.Ptr);
+                _currentEndPtr = _currentPtr + size;
+
+                // Copy the key mapping and content.
+                key._keyMappingDictionary.CopyTo(_keyMappingDictionary.AsSpan());
+                key._keyMappingIndex.CopyTo(_keyMappingIndex.AsSpan());
+                Memory.Copy(_storage.Ptr, key._storage.Ptr, size);
+            }
+
+            public bool IsValid => Dictionary > 0;
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            public ReadOnlySpan<byte> EncodedWithCurrent()
+            {
+                Debug.Assert(IsValid, "Cannot get an encoded key without a current dictionary");
+                return EncodedWith(Dictionary);
+            }
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            public byte* EncodedWith(long dictionaryId, out int length)
+            {
+                [SkipLocalsInit]
+                [MethodImpl(MethodImplOptions.NoInlining)]
+                byte* EncodeFromDecodedForm()
+                {
+                    Debug.Assert(IsValid, "At this stage we either created the key using an unencoded version. Current dictionary cannot be invalid.");
+
+                    // We acquire the decoded form. This will lazily evaluate if needed. 
+                    ReadOnlySpan<byte> decodedKey = Decoded();
+
+                    // IMPORTANT: Pointers are potentially invalidated by the grow storage call but not the indexes.
+
+                    // We look for an appropriate place to put this key
+                    int buckedIdx = SelectHashBucketForWrite();
```

Comment:
I though we aren't doing hashing any more?

Reply:
Need to change the name, yes.